source/robot_model/src/Model.cpp

```diff
@@ -260,9 +262,138 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
+    }
+    else{
+      W_b(n,n) = 1;
+    }      
+  }
+  
+  return W_b;
+}
+
+Eigen::VectorXd Model::cwln_repulsive_potential_field(const state_representation::JointPositions& joint_positions, const double& margin) {
+  Eigen::VectorXd Psi(this->robot_model_.nq);
+  Eigen::VectorXd q = joint_positions.get_positions();
+  for(int i=0; i < this->robot_model_.nq; i++){
+    Psi[i]=0;
+    if(q[i] < this->robot_model_.lowerPositionLimit[i] + margin){
+      Psi[i] = this->robot_model_.upperPositionLimit[i] - margin - std::max(q[i], this->robot_model_.lowerPositionLimit[i]);
+    }
+    else if(this->robot_model_.upperPositionLimit[i] - margin < q[i]){
+      Psi[i] = this->robot_model_.lowerPositionLimit[i] + margin - std::min(q[i], this->robot_model_.upperPositionLimit[i]);
+    }     
+  }
+  
+  return Psi;
+}
+
+void Model::clamp_joint_positions(state_representation::JointPositions& joint_positions) {
+  Eigen::VectorXd q_clamped = joint_positions.get_positions();
+  for(int i=0; i < this->robot_model_.nq; i++){
+    q_clamped[i] = std::min(std::max(q_clamped[i], this->robot_model_.lowerPositionLimit[i]), this->robot_model_.upperPositionLimit[i]);
+  }
+  joint_positions.set_positions(q_clamped);
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                            const state_representation::JointState& current_joint_state,
+                                                            std::string frame_name,
+                                                            Model::inverse_geometry_parameters params) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  }
+  else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+
+  state_representation::JointPositions q(current_joint_state);
+  state_representation::JointVelocities dq(this->get_robot_name(), this->get_number_of_joints());
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  const pinocchio::SE3 oMdes(desired_cartesian_state.get_orientation().toRotationMatrix(), desired_cartesian_state.get_position());
+  
+  // 1 second pour le Newton-Raphson algorithm
+  std::chrono::nanoseconds dt((int)(1e9));
+
+  pinocchio::Data::Matrix6x J(6,this->robot_model_.nq);
+  pinocchio::Data::Matrix6x J_b(6,this->robot_model_.nq);
+  J.setZero();
+  J_b.setZero();
+  Eigen::VectorXd v(this->robot_model_.nq);
+  Eigen::VectorXd delta_q(this->robot_model_.nq);
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::MatrixXd W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::VectorXd psi(this->robot_model_.nq);
+
+  this->forward_geometry(q, frame_id);
```

Comment:
`forward_geometry` returns the end-effector pose so if you are not storing the result this means this call is for nothing.

Reply:
ah actually this is probably because you count on pinocchio that actually updates the `robot_data_`. But this is a side effect. It is to me bad design on pinocchio side. If you need the pose of the end-effector then retrieve it and use it, or call the proper pinocchio function you need but don't rely on side effects from functions.

Reply:
yes exactly. I can also use the cartesian state coming from our forward geometry, but I don't know how exactly calculate the error in the lie group SE(3). can I just do: X_d - X?